File: app/src/main/kotlin/tech/mmarca/openvitals/ui/components/CrossMetricInsightCard.kt
package tech.mmarca.openvitals.ui.components

import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.outlined.TrendingDown
import androidx.compose.material.icons.automirrored.outlined.TrendingFlat
import androidx.compose.material.icons.automirrored.outlined.TrendingUp
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import tech.mmarca.openvitals.R
import tech.mmarca.openvitals.domain.insights.CrossMetricDirection
import tech.mmarca.openvitals.domain.insights.CrossMetricInsight
import tech.mmarca.openvitals.domain.insights.CrossMetricStrength
import kotlin.math.abs
import kotlin.math.roundToInt

@Composable
fun CrossMetricInsightCard(
    insight: CrossMetricInsight,
    title: String,
    positiveMessage: String,
    negativeMessage: String,
    neutralMessage: String,
    accentColor: Color,
    modifier: Modifier = Modifier,
) {
    val shape = MaterialTheme.shapes.medium
    val message = when {
        insight.strength == CrossMetricStrength.WEAK -> neutralMessage
        insight.direction == CrossMetricDirection.POSITIVE -> positiveMessage
        insight.direction == CrossMetricDirection.NEGATIVE -> negativeMessage
        else -> neutralMessage
    }
    val relationship = when {
        insight.strength == CrossMetricStrength.WEAK -> stringResource(R.string.cross_metric_weak_link)
        insight.direction == CrossMetricDirection.POSITIVE -> stringResource(R.string.cross_metric_positive_link)
        insight.direction == CrossMetricDirection.NEGATIVE -> stringResource(R.string.cross_metric_negative_link)
        else -> stringResource(R.string.cross_metric_weak_link)
    }

    OpenVitalsCard(
        modifier = modifier
            .fillMaxWidth()
            .border(
                width = 1.dp,
                color = MaterialTheme.colorScheme.outlineVariant.copy(alpha = 0.7f),
                shape = shape,
            ),
        shape = shape,
    ) {
        Column(modifier = Modifier.padding(16.dp)) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.Top,
            ) {
                Row(
                    modifier = Modifier.weight(1f),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    Icon(
                        imageVector = when {
                            insight.strength == CrossMetricStrength.WEAK -> Icons.AutoMirrored.Outlined.TrendingFlat
                            insight.direction == CrossMetricDirection.POSITIVE -> Icons.AutoMirrored.Outlined.TrendingUp
                            insight.direction == CrossMetricDirection.NEGATIVE -> Icons.AutoMirrored.Outlined.TrendingDown
                            else -> Icons.AutoMirrored.Outlined.TrendingFlat
                        },
                        contentDescription = null,
                        tint = accentColor,
                        modifier = Modifier.size(24.dp),
                    )
                    Spacer(Modifier.width(12.dp))
                    Column {
                        Text(
                            text = title,
                            style = MaterialTheme.typography.titleSmall,
                            fontWeight = FontWeight.SemiBold,
                        )
                        Text(
                            text = relationship,
                            style = MaterialTheme.typography.bodySmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                        )
                    }
                }
                val correlationPercent = (insight.correlation * 100.0).roundToInt()
                Text(
                    text = stringResource(
                        R.string.cross_metric_correlation,
                        signedPercent(correlationPercent),
                    ),
                    style = MaterialTheme.typography.titleMedium,
                    fontWeight = FontWeight.Bold,
                    color = MaterialTheme.colorScheme.onSurface,
                )
            }
            Spacer(Modifier.height(12.dp))
            Text(
                text = message,
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
            Spacer(Modifier.height(8.dp))
            Text(
                text = stringResource(R.string.cross_metric_paired_days, insight.pairedDays),
                style = MaterialTheme.typography.labelMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
    }
}

private fun signedPercent(value: Int): String =
    when {
        value > 0 -> "+${abs(value)}%"
        value < 0 -> "-${abs(value)}%"
        else -> "0%"
    }

File: lib/ui/components/cross_metric_insight_card.dart
import 'package:flutter/material.dart';

import '../../domain/insights/cross_metric_insights.dart';
import '../../l10n/app_localizations.dart';
import 'ov_card.dart';

/// Port of the Kotlin `CrossMetricInsightCard`: how two metrics moved together
/// over the period, as a correlation with a plain-language reading.

/// Kotlin `signedPercent`.
String signedPercent(int value) {
  if (value > 0) return '+${value.abs()}%';
  if (value < 0) return '-${value.abs()}%';
  return '0%';
}

class CrossMetricInsightCard extends StatelessWidget {
  const CrossMetricInsightCard({
    super.key,
    required this.insight,
    required this.title,
    required this.positiveMessage,
    required this.negativeMessage,
    required this.neutralMessage,
    required this.accentColor,
  });

  final CrossMetricInsight insight;
  final String title;
  final String positiveMessage;
  final String negativeMessage;
  final String neutralMessage;
  final Color accentColor;

  /// A weak correlation reads as "no clear pattern" whatever its sign — the
  /// direction is only trusted once the link is strong enough.
  bool get _isWeak => insight.strength == CrossMetricStrength.weak;

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    final l10n = AppLocalizations.of(context);

    final message = _isWeak
        ? neutralMessage
        : switch (insight.direction) {
            CrossMetricDirection.positive => positiveMessage,
            CrossMetricDirection.negative => negativeMessage,
            CrossMetricDirection.neutral => neutralMessage,
          };
    final relationship = _isWeak
        ? l10n.crossMetricWeakLink
        : switch (insight.direction) {
            CrossMetricDirection.positive => l10n.crossMetricPositiveLink,
            CrossMetricDirection.negative => l10n.crossMetricNegativeLink,
            CrossMetricDirection.neutral => l10n.crossMetricWeakLink,
          };
    final icon = _isWeak
        ? Icons.trending_flat
        : switch (insight.direction) {
            CrossMetricDirection.positive => Icons.trending_up,
            CrossMetricDirection.negative => Icons.trending_down,
            CrossMetricDirection.neutral => Icons.trending_flat,
          };

    return DecoratedBox(
      decoration: BoxDecoration(
        borderRadius: const BorderRadius.all(Radius.circular(12)),
        border: Border.all(
          color: theme.colorScheme.outlineVariant.withValues(alpha: 0.7),
        ),
      ),
      child: OpenVitalsCard(
        child: Padding(
          padding: const EdgeInsets.all(16),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Row(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Expanded(
                    child: Row(
                      children: [
                        Icon(icon, size: 24, color: accentColor),
                        const SizedBox(width: 12),
                        Expanded(
                          child: Column(
                            crossAxisAlignment: CrossAxisAlignment.start,
                            children: [
                              Text(
                                title,
                                style: theme.textTheme.titleSmall
                                    ?.copyWith(fontWeight: FontWeight.w600),
                              ),
                              Text(
                                relationship,
                                style: theme.textTheme.bodySmall?.copyWith(
                                    color: theme.colorScheme.onSurfaceVariant),
                              ),
                            ],
                          ),
                        ),
                      ],
                    ),
                  ),
                  Text(
                    l10n.crossMetricCorrelation(
                      signedPercent((insight.correlation * 100.0).round()),
                    ),
                    style: theme.textTheme.titleMedium?.copyWith(
                      fontWeight: FontWeight.bold,
                      color: theme.colorScheme.onSurface,
                    ),
                  ),
                ],
              ),
              const SizedBox(height: 12),
              Text(
                message,
                style: theme.textTheme.bodyMedium
                    ?.copyWith(color: theme.colorScheme.onSurfaceVariant),
              ),
              const SizedBox(height: 8),
              Text(
                l10n.crossMetricPairedDays(insight.pairedDays),
                style: theme.textTheme.labelMedium
                    ?.copyWith(color: theme.colorScheme.onSurfaceVariant),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
